pkgs/native_assets_cli/lib/src/model/asset.dart

```diff
@@ -0,0 +1,323 @@
+
+  Asset({
+    required this.name,
+    required this.packaging,
+    required this.target,
+    required this.path,
+  });
+
+  factory Asset.fromYaml(YamlMap yamlMap) => Asset(
+        name: yamlMap[_nameKey] as String,
+        path: AssetPath.fromYaml(yamlMap[_pathKey] as YamlMap),
+        target: Target.fromString(yamlMap[_targetKey] as String),
+        packaging: Packaging.fromName(yamlMap[_packagingKey] as String),
+      );
+
+  static List<Asset> listFromYamlString(String yaml) {
+    final yamlObject = loadYaml(yaml);
+    if (yamlObject == null) {
+      return [];
+    }
+    return [
+      for (final yamlElement in yamlObject as YamlList)
+        Asset.fromYaml(yamlElement as YamlMap),
+    ];
+  }
+
+  static List<Asset> listFromYamlList(YamlList yamlList) => [
+        for (final yamlElement in yamlList)
+          Asset.fromYaml(yamlElement as YamlMap),
+      ];
+
+  Asset copyWith({
+    Packaging? packaging,
+    String? name,
+    Target? target,
+    AssetPath? path,
+  }) =>
+      Asset(
+        name: name ?? this.name,
+        packaging: packaging ?? this.packaging,
+        target: target ?? this.target,
+        path: path ?? this.path,
+      );
+
+  @override
+  bool operator ==(Object other) {
+    if (other is! Asset) {
+      return false;
+    }
+    return other.name == name &&
+        other.packaging == packaging &&
+        other.target == target &&
+        other.path == path;
+  }
+
+  @override
+  int get hashCode => Object.hash(name, packaging, target, path);
+
+  Map<String, Object> toYaml() => {
+        _nameKey: name,
+        _packagingKey: packaging.name,
+        _pathKey: path.toYaml(),
+        _targetKey: target.toString(),
+      };
+
+  Map<String, List<String>> toDartConst() => {
+        name: path.toDartConst(),
+      };
+
+  String toYamlString() => yamlEncode(toYaml());
+
+  static const _nameKey = 'name';
+  static const _packagingKey = 'packaging';
+  static const _pathKey = 'path';
+  static const _targetKey = 'target';
+
+  Future<bool> exists() => path.exists();
+
+  @override
+  String toString() => 'Asset(${toYaml()})';
+}
+
+extension AssetIterable on Iterable<Asset> {
+  List<Object> toYaml() => [for (final item in this) item.toYaml()];
+
+  String toYamlString() => yamlEncode(toYaml());
+
+  Iterable<Asset> wherePackaging(Packaging packaging) =>
+      where((e) => e.packaging == packaging);
+
+  Map<Target, List<Asset>> get assetsPerTarget {
+    final result = <Target, List<Asset>>{};
+    for (final asset in this) {
+      final assets = result[asset.target] ?? [];
+      assets.add(asset);
+      result[asset.target] = assets;
+    }
+    return result;
+  }
+
+  Map<String, Map<String, List<String>>> toDartConst() => {
+        for (final entry in assetsPerTarget.entries)
+          entry.key.toString():
+              _combineMaps(entry.value.map((e) => e.toDartConst()).toList())
+      };
+
+  Map<Object, Object> toNativeAssetsFileEncoding() => {
```

Comment:
I was thinking that it could be conceivable that people end up compiling kernel themselves and want to provide a manually baked `native_assets.yaml` similiar to how they can provide a manual `package_config.json`.

We could move the implementation to pkg/native_assets_builder in https://dart-review.googlesource.com/c/sdk/+/267340 if we want it to be private.